units/en/unit2/smolagents/tools.mdx

```diff
@@ -0,0 +1,267 @@
+
+<Tip>
+You can follow the code in <a href="https://huggingface.co/agents-course/notebooks/blob/main/unit2/smolagents/tools.ipynb" target="_blank">this notebook</a> that you can run using Google Colab.
+</Tip>
+
+Let's imagine that Alfred has already decided on the menu for the party, but now he needs help preparing food for such a large number of guests. To do so, he would like to hire a catering service and needs to identify the highest-rated options available. Alfred can leverage a tool to search for the best catering services in his area. 
+
+Below is an example of how Alfred can use the `@tool` decorator to make this happen:
+
+```python
+from smolagents import CodeAgent, HfApiModel, tool
+
+# Let's pretend we have a function that fetches the highest-rated catering services.
+@tool
+def catering_service_tool(query: str) -> str:
+    """
+    This tool returns the highest-rated catering service in Gotham City.
+    
+    Args:
+        query: A search term for finding catering services.
+    """
+    # Example list of catering services and their ratings
+    services = {
+        "Gotham Catering Co.": 4.9,
+        "Wayne Manor Catering": 4.8,
+        "Gotham City Events": 4.7,
+    }
+    
+    # Find the highest rated catering service (simulating search query filtering)
+    best_service = max(services, key=services.get)
+    
+    return best_service
+
+
+agent = CodeAgent(tools=[catering_service_tool], model=HfApiModel())
+
+# Run the agent to find the best catering service
+result = agent.run(
+    "Can you give me the name of the highest-rated catering service in Gotham City?"
+)
+
+print(result)   # Output: Gotham Catering Co.
+```
+
+### Defining a Tool as a Python Class  
+
+This approach involves creating a subclass of [`Tool`](https://huggingface.co/docs/smolagents/v1.8.1/en/reference/tools#smolagents.Tool).  For complex tools, we can implement a class instead of a Python function. The class wraps the function with metadata that helps the LLM understand how to use it effectively. In this class, we define:  
+
+- `name`: The tool's name.  
+- `description`: A description used to populate the agent's system prompt.  
+- `inputs`: A dictionary with keys `type` and `description`, providing information to help the Python interpreter process inputs.  
+- `output_type`: Specifies the expected output type.  
+- `forward`: The method containing the inference logic to execute.
+
+Below, we can see an example of a tool built using `Tool` and to integrate it within a `CodeAgent`.
+
+#### Generating a tool to generate ideas about the superhero-themed party
+
+Alfred's party at the mansion is a **superhero-themed event**, but he needs some creative ideas to make it truly special. As a fantastic host, he wants to surprise the guests with a unique theme. 
+
+To do this, he can use an agent that generates superhero-themed party ideas based on a given category. This way, Alfred can find the perfect party theme to wow his guests.
+
+```python
```

Comment:
It would be nice to show some state to demonstrate the benefits of using a class vs a single function.